game/scripts/vscripts/abilities/stopfightingyourself/dupe_heroes.lua

```diff
@@ -0,0 +1,145 @@
+    FIND_ANY_ORDER,
+    false
+  )
+
+
+  for _,unit in ipairs(units) do
+    if unit:IsRealHero() and unit ~= caster then
+      self:StartCooldown(self:GetSpecialValueFor('cooldown'))
+
+      local illusion = CreateUnitByName(
+        unit:GetName(),
+        unit:GetAbsOrigin(),
+        true,
+        caster,
+        caster,
+        --nil,
+        --nil,
+        caster:GetTeamNumber()
+      )
+
+      -- Level
+      local target_level = unit:GetLevel()
+      for i = 1, target_level - 1 do
+        illusion:HeroLevelUp(false)
+        HeroProgression:ReduceStatGain(illusion, i)
+      end
+
+      illusion:SetAbilityPoints(0)
+
+      -- Abilities
+      for slot = 0, 15 do
+        local theirAbility = unit:GetAbilityByIndex(slot)
+
+        if theirAbility then
+          local ourAbility = illusion:FindAbilityByName(theirAbility:GetAbilityName())
+          if ourAbility then
+            ourAbility:SetLevel(theirAbility:GetLevel())
+          end
+        end
+      end
+
+
+      -- Items
+      for slot = DOTA_ITEM_SLOT_1, DOTA_ITEM_SLOT_6 do
+        local theirItem = unit:GetItemInSlot(slot)
+
+        if theirItem then
+          local ourItem = illusion:AddItemByName(theirItem:GetAbilityName())
+
+          if ourItem:RequiresCharges() then
+            local charges = theirItem:GetCurrentCharges()
+            ourItem:SetCurrentCharges(charges)
+          end
+        end
+      end
+
+
+      -- Illusion Modifier
+      illusion:AddNewModifier(
+        caster,
+        self,
+        "modifier_illusion",
+        {
+          duration = -1,
+          outgoing_damage = self:GetSpecialValueFor('illusion_outgoing_damage'),
+          incoming_damage = self:GetSpecialValueFor('illusion_incoming_damage'),
+        }
+      )
+
+      --illusion:MakeIllusion()
+
+      -- Stats
+      -- BUG Gold bounty does not work
+      -- NOTE XP probably too
+      illusion:SetMaximumGoldBounty(0)
+      illusion:SetMinimumGoldBounty(0)
+      illusion:SetDeathXP(0)
+      illusion:SetRespawnsDisabled(true)
+
+      -- Set Stats to original hero
+      illusion:SetHealth(unit:GetHealth())
+      illusion:SetMana(unit:GetMana())
+
+      -- Randomly play sound to player (10% chance)
+      -- NOTE this doesn't seem to work
+      if math.random(100) <= 10 then
+        EmitAnnouncerSoundForPlayer('sounds/vo/announcer_dlc_rick_and_morty/generic_illusion_based_hero_02.vsnd', unit:GetPlayerID())
+      end
+
+      if caster.illusions == nil then
+        caster.illusions = {}
+      end
+      local index = #caster.illusions + 1
+      caster.illusions[index] = illusion
+
+      illusion:OnDeath(function()
+        -- create particle
+        -- NOTE I have yet to see those particles
+        Timers:CreateTimer(0.8, function()
+          ParticleManager:CreateParticle('particles/generic_gameplay/illusion_killed.vpcf', PATTACH_ABSORIGIN, illusion)
+        end)
+        illusion:RemoveSelf()
+        table.remove(caster.illusions, index)
```

Comment:
Just realised that this would actually break pretty easily as well, since the indices won't be fixed. Possible solutions include:
- Searching for the correct illusion instance to delete using `index` (this could possibly result in rare race conditions if illusions die at the same time)
- Doing garbage collection in the boss thinker instead by deleting illusions that are no longer alive
- Going back to what you had before and using `pairs` to loop through the illusions instead. Not sure if this has any other potential issues